modules/priceFloors.js

```diff
@@ -0,0 +1,643 @@
+      fetching = true;
+    }
+  } else if (fetching) {
+    utils.logWarn(`${MODULE_NAME}: A fetch is already occuring. Skipping.`);
+  }
+}
+
+/**
+ * @summary Updates our allowedFields and fieldMatchingFunctions with the publisher defined new ones
+ */
+function addFieldOverrides(overrides) {
+  Object.keys(overrides).forEach(override => {
+    // we only add it if it is not already in the allowed fields and if the passed in value is a function
+    if (allowedFields.indexOf(override) === -1 && typeof overrides[override] === 'function') {
+      allowedFields.push(override);
+      fieldMatchingFunctions[override] = overrides[override];
+    }
+  });
+}
+
+/**
+ * @summary This is the function which controls what happens during a pbjs.setConfig({...floors: {}}) is called
+ */
+export function handleSetFloorsConfig(config) {
+  _floorsConfig = utils.pick(config, [
+    'enabled', enabled => enabled !== false, // defaults to true
+    'auctionDelay', auctionDelay => auctionDelay || 0,
+    'endpoint', endpoint => endpoint || {},
+    'enforcement', enforcement => utils.pick(enforcement || {}, [
+      'enforceJS', enforceJS => enforceJS !== false, // defaults to true
+      'enforcePBS', enforcePBS => enforcePBS === true, // defaults to false
+      'floorDeals', floorDeals => floorDeals === true, // defaults to false
+      'bidAdjustment', bidAdjustment => bidAdjustment !== false, // defaults to true
+    ]),
+    'additionalSchemaFields', additionalSchemaFields => typeof additionalSchemaFields === 'object' && Object.keys(additionalSchemaFields).length > 0 ? addFieldOverrides(additionalSchemaFields) : undefined,
+    'data', data => (data && parseFloorData(data, 'setConfig')) || _floorsConfig.data // do not overwrite if passed in data not valid
+  ]);
+
+  // if enabled then do some stuff
+  if (_floorsConfig.enabled) {
+    // handle the floors fetch
+    generateAndHandleFetch(_floorsConfig.endpoint);
+
+    if (!addedFloorsHook) {
+      // register hooks / listening events
+      // when auction finishes remove it's associated floor data
+      events.on(CONSTANTS.EVENTS.AUCTION_END, (args) => delete _floorDataForAuction[args.auctionId]);
+
+      // we want our hooks to run after the currency hooks
+      getGlobal().requestBids.before(requestBidsHook, 50);
+      // if user has debug on then we want to allow the debugging module to run before this, assuming they are testing priceFloors
+      // debugging is currently set at 5 priority
+      getHook('addBidResponse').before(addBidResponseHook, utils.debugTurnedOn() ? 4 : 50);
+      addedFloorsHook = true;
+    }
+  } else {
+    utils.logInfo(`${MODULE_NAME}: Turning off module`);
+
+    _floorsConfig = {};
+    _floorDataForAuction = {};
+
+    getHook('addBidResponse').getHooks({hook: addBidResponseHook}).remove();
+    getGlobal().requestBids.getHooks({hook: requestBidsHook}).remove();
+
+    addedFloorsHook = false;
```

Comment:
Can you clarify the use-case for this else clause?  

From my understanding, this would only run if someone set up a setConfig with the priceFloors config defined, but set the `enabled` field to `false`.  But if they were to do that, then wouldn't it be likely that none of these items were set up in the first place?  Wouldn't someone have to enable the config during one auction, then disabled on the same page session for a subsequent auction (like a refresh) for this code to reset stuff (at least the hooks)?

Is there some other scenario in mind that this code should handle?

Reply:
Yeah, if for a given page load they enabled floors and wanted to disable it then this is the use case.

I am not too sure on if in real life this is something which will happen, but for example if you have a prebid bundle which by default enables floors but you wanted to turn it off on a subset of your sites you could do so on those specific pages.

Not something I think we for sure need, so if you want I am open to changing the behavior.

Reply:
No, it makes sense.  This doesn't appear to harm anything if someone were to have it configured to off by default, so we can leave it in just in case.